// @ar.io/sdk/lib/cjs/types/io.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isLeasedArNSRecord = exports.isProcessIdConfiguration = exports.isProcessConfiguration = exports.isValidIntent = exports.intentsUsingYears = exports.validIntents = void 0;
const arweave_js_1 = require("../utils/arweave.js");
exports.validIntents = [
    'Buy-Record',
    'Extend-Lease',
    'Increase-Undername-Limit',
    'Upgrade-Name',
    'Primary-Name-Request',
];
exports.intentsUsingYears = ['Buy-Record', 'Extend-Lease'];
const isValidIntent = (intent) => {
    return exports.validIntents.indexOf(intent) !== -1;
};
exports.isValidIntent = isValidIntent;
// Typeguard functions
function isProcessConfiguration(config) {
    return 'process' in config;
}
exports.isProcessConfiguration = isProcessConfiguration;
function isProcessIdConfiguration(config) {
    return ('processId' in config &&
        typeof config.processId === 'string' &&
        (0, arweave_js_1.validateArweaveId)(config.processId) === true);
}
exports.isProcessIdConfiguration = isProcessIdConfiguration;
function isLeasedArNSRecord(record) {
    return record.type === 'lease';
}
exports.isLeasedArNSRecord = isLeasedArNSRecord;
